File: python/525.py
"""Project Euler Problem 525: Rolling Ellipse.

Let C(a, b) be the length of the path traced by the center of the ellipse
x²/a² + (y-b)²/b² = 1 as it rolls one full revolution along the x-axis.
Find C(1, 4) + C(3, 4).
"""

from __future__ import annotations

import math


def hypot(x: float, y: float) -> float:
    """Hypotenuse."""
    return math.sqrt(x * x + y * y)


def solve() -> str:
    """Solve Problem 525."""
    L = 1000000

    def C(a: float, b: float) -> float:
        """Compute C(a, b)."""
        prev_perim_x = 0.0
        prev_perim_y = b
        prev_x = 0.0
        prev_y = b
        C_val = 0.0

        for i in range(1, L + 1):
            theta = math.pi / 2 / L * i
            sin_val = math.sin(theta)
            cos_val = math.cos(theta)
            perim_x = a * sin_val
            perim_y = b * cos_val
            r = hypot(perim_x, perim_y)
            alpha = math.atan(perim_x / perim_y) + math.atan(
                (a * cos_val) / (b * sin_val)
            )
            x = r * math.cos(alpha)
            y = r * math.sin(alpha)
            C_val += hypot(
                x - prev_x + hypot(perim_x - prev_perim_x, perim_y - prev_perim_y),
                y - prev_y,
            )
            prev_perim_x = perim_x
            prev_perim_y = perim_y
            prev_x = x
            prev_y = y

        return 4 * C_val

    ans = C(1, 4) + C(3, 4)
    return f"{ans:.8f}"


def main() -> str:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/525.c
/*
 * Project Euler 525 - Rolling Ellipse
 *
 * Compute C(a,b) = length of path traced by center of ellipse
 * x^2/a^2 + (y-b)^2/b^2 = 1 as it rolls along the x-axis.
 * Answer: C(1,4) + C(3,4).
 */
#include <stdio.h>
#include <math.h>

#define L 1000000

static double C(double a, double b) {
    double prev_perim_x = 0.0;
    double prev_perim_y = b;
    double prev_x = 0.0;
    double prev_y = b;
    double c_val = 0.0;

    double step = M_PI / 2.0 / L;

    for (int i = 1; i <= L; i++) {
        double theta = step * i;
        double sin_val = sin(theta);
        double cos_val = cos(theta);
        double perim_x = a * sin_val;
        double perim_y = b * cos_val;
        double r = hypot(perim_x, perim_y);
        double alpha = atan(perim_x / perim_y) + atan((a * cos_val) / (b * sin_val));
        double x = r * cos(alpha);
        double y = r * sin(alpha);

        double dx_perim = perim_x - prev_perim_x;
        double dy_perim = perim_y - prev_perim_y;
        double arc = hypot(dx_perim, dy_perim);

        c_val += hypot(x - prev_x + arc, y - prev_y);

        prev_perim_x = perim_x;
        prev_perim_y = perim_y;
        prev_x = x;
        prev_y = y;
    }

    return 4.0 * c_val;
}

int main(void) {
    double ans = C(1.0, 4.0) + C(3.0, 4.0);
    printf("%.8f\n", ans);
    return 0;
}
